mypy/semanal.py

```diff
@@ -3230,10 +3237,16 @@ def name_not_defined(self, name: str, ctx: Context) -> None:
                 self.add_fixture_note(fullname, ctx)
 
     def name_already_defined(self, name: str, ctx: Context,
-                             original_ctx: Optional[SymbolTableNode] = None) -> None:
+                             original_ctx: Optional[Union[SymbolTableNode, SymbolNode]] = None) -> None:
```

Comment:
In general, it is not a good idea to use a union here, but it seems to me the old signature was "wrong", so it is fine in this case. Maybe you can add a short comment about this?

Reply:
The old signature was actually correct, the function used to only be called with `SymbolTableNode`s. I added the option for passing a `SymbolNode` instead because in `_visit_overloaded_func_def` I only have a `SymbolNode` to pass it, not a `SymbolTableNode`.

Reply:
So I guess to fix this, the options are
- take `Optional[SymbolNode]` instead
- take `Optiona[SymbolTableNode]` only, and in `_visit_overloaded_func_def` create a `SymbolTableNode` to pass to the function

Reply:
I mean that `SymbolTableNode` is logically "wrong" because it is just an entry in the symbol table wrapper, while `SymbolNode` is an actual error context (well to be more precise, `SymbolNode`s are semantic objects, not syntactic, but at least they are closer, and should ideally point to the line of syntax node defining them). So I vote for `SymbolNode`.

Reply:
Actually you might need `.kind == MODULE_REF` which is only available on `SymbolTableNode`. Potentially, this can be replaced by `isinstance(node, (MypyFile, ImportedName))`, but I think it is better to just leave it as is.

Reply:
Okay, I'll leave it as-is then.